File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast = 0
        slow = 0
        
        while True:
            fast = nums[nums[fast]]
            slow = nums[slow]
            
            if fast == slow:
                pointer = 0
                
                while pointer != slow:
                    pointer = nums[pointer]
                    slow = nums[slow]
                return pointer
                    
        

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function(nums) {
    //singly linked cycle
    let fast = 0;
    let slow = 0;
    
    while (true) {
        fast = nums[nums[fast]]; //increment fast by 2
        slow = nums[slow]; //increment slow by 1
        
        if (fast===slow) {
            let pointer = 0; //another slow pointer
            
            while (pointer !== slow) {
                pointer = nums[pointer]; //increment pointer by 1
                slow = nums[slow]; //increment slow by 1
            }
            return pointer;
        }
    }
};